presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkAdaptiveQueryExecution.java

```diff
@@ -0,0 +1,351 @@
+            JsonCodec<PrestoSparkQueryData> queryDataJsonCodec,
+            PrestoSparkRddFactory rddFactory,
+            TransactionManager transactionManager,
+            PagesSerde pagesSerde,
+            PrestoSparkExecutionExceptionFactory executionExceptionFactory,
+            Duration queryTimeout,
+            long queryCompletionDeadline,
+            PrestoSparkMetadataStorage metadataStorage,
+            Optional<String> queryStatusInfoOutputLocation,
+            Optional<String> queryDataOutputLocation,
+            TempStorage tempStorage,
+            NodeMemoryConfig nodeMemoryConfig,
+            Set<PrestoSparkServiceWaitTimeMetrics> waitTimeMetrics,
+            Optional<ErrorClassifier> errorClassifier,
+            PrestoSparkPlanFragmenter planFragmenter,
+            Metadata metadata,
+            PartitioningProviderManager partitioningProviderManager,
+            HistoryBasedPlanStatisticsTracker historyBasedPlanStatisticsTracker)
+    {
+        super(
+                sparkContext,
+                session,
+                queryMonitor,
+                taskInfoCollector,
+                shuffleStatsCollector,
+                taskExecutorFactory,
+                taskExecutorFactoryProvider,
+                queryStateTimer,
+                warningCollector,
+                query,
+                planAndMore,
+                sparkQueueName,
+                taskInfoCodec,
+                sparkTaskDescriptorJsonCodec,
+                queryStatusInfoJsonCodec,
+                queryDataJsonCodec,
+                rddFactory,
+                transactionManager,
+                pagesSerde,
+                executionExceptionFactory,
+                queryTimeout,
+                queryCompletionDeadline,
+                metadataStorage,
+                queryStatusInfoOutputLocation,
+                queryDataOutputLocation,
+                tempStorage,
+                nodeMemoryConfig,
+                waitTimeMetrics,
+                errorClassifier,
+                planFragmenter,
+                metadata,
+                partitioningProviderManager,
+                historyBasedPlanStatisticsTracker);
+
+        this.executedFragments = new HashSet<>();
+        this.iterativePlanFragmenter = createIterativePlanFragmenter();
+        this.fragmentEventQueue = new LinkedBlockingQueue<>();
+    }
+
+    private IterativePlanFragmenter createIterativePlanFragmenter() {
+        // Function<PlanFragmentId, SubPlan> getSubPlanForId = null;
+
+        // TODO Rebecca to check: I have not found any place that this is being called with true but let me know.
+        boolean forceSingleNode = false;
+        Function<PlanFragmentId, Boolean> isFragmentFinished = this.executedFragments::contains;
+
+        // TODO The static plan fragmenter is created by injection (and is also passed in the constructor here but remains unused).
```

Comment:
Actually, I think it wouldn't be too difficult to make IterativePlanFragmenter stateless.  I can do that after I finish doing the stats/optimizer stuff and then we can inject it.

Reply:
did it quickly just now. let me know what you think: https://github.com/prestodb/presto/pull/18811.  It means you'll have to pass more things into the createReadySubPlans call instead of passing into the IterativePlanFragmenter constructor.  also, you'll need to keep track of the fragmentId -> subPlan map and pass a function we can use to get that info (similar to isFragmentReady)

Reply:
Left a TODO so that we address this once we merge 18811.